In [1]:
import pandas as pd
import numpy as np

# visualization tools
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.graph_objects as go
from plotly.subplots import make_subplots

# kaggle utils
import kaggle_utils_py as kaggle_utils


# <p style="background-color:#9900cc;padding:15px;font-family:newtimeroman;color:#ffff80;font-size:110%;border-radius:40px 5px;"> 1 | Basic stuff</p>


In [51]:
data = pd.read_csv("/home/pavithra/projects/Kaggle/EDA&Hypothesis_testing_master_degree_program/data/201709301651_masters_portal.csv")
data.head()

country_name country_code                 university_name  university_rank  \
0      Armenia          ARM  American University of Armenia              NaN   
1      Armenia          ARM  American University of Armenia              NaN   
2      Armenia          ARM  American University of Armenia              NaN   
3      Armenia          ARM  American University of Armenia              NaN   
4      Armenia          ARM  American University of Armenia              NaN   

                                    program_name program_type  \
0                                      Economics          MSc   
1    Political Science and International Affairs       Master   
2                        Business Administration          MBA   
3               Computer and Information Science          MSc   
4  Industrial Engineering and Systems Management         MEng   

               deadline   duration language tution_1_currency  ...  \
0  2004-07-18T00:00:00Z        NaN  English               EUR  ...   
1  2031-07-18T00:00:00Z  24 months  English               EUR  ...   
2  2004-07-18T00:00:00Z        NaN  English               EUR  ...   
3                   NaN  24 months  English               EUR  ...   
4  2031-07-18T00:00:00Z  24 months  English               EUR  ...   

   tution_2_money tution_2_type tuition_price_specification  \
0          2108.0        EU/EEA              Tuition (Year)   
1          2500.0      National              Tuition (Year)   
2          2499.0        EU/EEA              Tuition (Year)   
3          2500.0      National              Tuition (Year)   
4          2500.0      National              Tuition (Year)   

              start_date ielts_score  \
0    2018-09-01 00:00:00         6.5   
1    2018-08-22 00:00:00         6.5   
2    2018-09-01 00:00:00         6.5   
3                    NaN         6.5   
4    2018-08-22 00:00:00         6.5   

                                           structure  \
0  ['Quantitative Methods for Economists (Mathema...   
1                                                NaN   
2  ['Managers with practical knowledge of account...   
3  ['Introduction to Object-Oriented Programming'...   
4  ['Probability Theory', 'Analysis and Design of...   

                                        academic_req  \
0  <section id="AcademicRequirements"> <h2>Academ...   
1  <section id="AcademicRequirements"> <h2>Academ...   
2  <section id="AcademicRequirements"> <h2>Academ...   
3  <section id="AcademicRequirements"> <h2>Academ...   
4  <section id="AcademicRequirements"> <h2>Academ...   

                                               facts         city  \
0  ['Starting in   2018-09-01 00:00:00    You can...  ['Yerevan']   
1  ['Starting in   2018-08-22 00:00:00    You can...  ['Yerevan']   
2  ['Starting in   2018-09-01 00:00:00    You can...  ['Yerevan']   
3  ['Deadline and start date Application deadline...  ['Yerevan']   
4  ['Starting in   2018-08-22 00:00:00    You can...  ['Yerevan']   

                                         program_url  
0  http://www.mastersportal.eu/studies/71101/econ...  
1  http://www.mastersportal.eu/studies/71085/poli...  
2  http://www.mastersportal.eu/studies/71102/busi...  
3  http://www.mastersportal.eu/studies/71104/comp...  
4  http://www.mastersportal.eu/studies/71103/indu...  

[5 rows x 23 columns]

In [52]:
data["academic_req"][0]

# We are not doing any kind of NLP here, and wanna check the price with other factors so this is not needed.

'<section id="AcademicRequirements"> <h2>Academic Requirements</h2> <section id="StudyRequirement"> <h6> Application Requirements</h6><p>To ensure your application for admission is complete, you must submit all of the required documents. Some programs have additional supplements. You should check the website of your preferred field of study to clarify whether there are additional admissions requirements. General application requirements include:</p><ul><li>A complete Application for Admission to Graduate Studyincluding all required supplements as outlined in the application instructions.</li></ul><ul><li>Present official results confirming English Language Proficiency typically through the TOEFL iBT or IELTS. Native and near native speakers are eligible for a waiver. For more information, please read the English proficiency waiver policy and application form here.<ul><li>You must submit official score reports confirming English language proficiency either from the TOEFL iBT or IELTS ex

In [53]:
# Remove the unwanted columns:
unwanted_columns = ["country_code", "deadline", "start_date", "structure", "academic_req", "facts", "program_url"]
data.drop(unwanted_columns, axis=1, inplace=True)

In [54]:
print("Total number of columns: {}".format(len(data.columns)))

print("Total number of rows: {}".format(len(data)))

print("--"*50)
# find the categorical columns
categorical_columns = data.select_dtypes(include=['object']).columns
continuous_columns = data.select_dtypes(include=["float64", "int64"]).columns
print("categorical Columns: {}".format(len(categorical_columns)))
print(categorical_columns)
print("continuous Columns: {}".format(len(continuous_columns)))
print(continuous_columns)

Total number of columns: 16
Total number of rows: 60425
----------------------------------------------------------------------------------------------------
categorical Columns: 12
Index(['country_name', 'university_name', 'program_name', 'program_type',
       'duration', 'language', 'tution_1_currency', 'tution_1_type',
       'tution_2_currency', 'tution_2_type', 'tuition_price_specification',
       'city'],
      dtype='object')
continuous Columns: 4
Index(['university_rank', 'tution_1_money', 'tution_2_money', 'ielts_score'], dtype='object')


In [55]:
data.describe()


university_rank  tution_1_money  tution_2_money   ielts_score
count     34901.000000    53773.000000    48086.000000  47357.000000
mean        278.745165    13956.115578     9030.554943      6.486992
std         202.120986    12908.596010    11581.640846      0.420575
min           1.000000        0.000000        0.000000      3.500000
25%         113.000000     2153.000000     1195.000000      6.500000
50%         251.000000    12850.000000     6120.000000      6.500000
75%         401.000000    19000.000000    10480.000000      6.500000
max         801.000000    99960.000000    99750.000000      9.000000

In [56]:
# How many levels present in the categorical columns.
for col in categorical_columns:
    print(f"The unique values in the '{col}' columns is ---- > {data[col].nunique()}")

The unique values in the 'country_name' columns is ---- > 94
The unique values in the 'university_name' columns is ---- > 2273
The unique values in the 'program_name' columns is ---- > 25352
The unique values in the 'program_type' columns is ---- > 502
The unique values in the 'duration' columns is ---- > 80
The unique values in the 'language' columns is ---- > 75
The unique values in the 'tution_1_currency' columns is ---- > 6
The unique values in the 'tution_1_type' columns is ---- > 4
The unique values in the 'tution_2_currency' columns is ---- > 6
The unique values in the 'tution_2_type' columns is ---- > 3
The unique values in the 'tuition_price_specification' columns is ---- > 8
The unique values in the 'city' columns is ---- > 1464


# My main goal is to analysis the relations between country, university, program and price

- 1 - which country gives the most master program
- 2 - which university gives the most master program
- 3 - which university has more price 
- 4 - which university has more number of programs
- 5 - does city has any effects on the price from the same university
- 6 - 

In [65]:
# 1 |  which country gives the most master program
program_data = data.groupby("country_name")["program_name"].count()

In [76]:
program_data = program_data.sort_values(ascending=False)
df = pd.DataFrame({"country_name": program_data.index,
                   "program_count": program_data.values})
df

country_name  program_count
0            United States          24084
1           United Kingdom          14130
2                Australia           2932
3                   Canada           2152
4              Netherlands           1346
..                     ...            ...
89               Guatemala              2
90                  Guyana              1
91                    Fiji              1
92                Ethiopia              1
93  Bosnia and Herzegovina              1

[94 rows x 2 columns]

In [132]:
# I can have a histgram for this
fig = go.Figure()

fig.add_trace(
    go.Bar(x=df["country_name"], y = df["program_count"], marker_color= '#9900cc')
)

fig.update_layout(
    template="plotly_dark",
    title="The number of master programs provided by each countries.",
    yaxis_title="Number of programs",
    bargap=0.2,

)

In [111]:
# Analysis:
# 1 | How many countries provides only one program ?
print("--"*50)
print("How many countries provides only one program ? ")
print(df[df["program_count"] == 1])

print("--"*50)
print("2 | How many programs provides by the top 5 countries?")
print(df.head())

print(df.iloc[0:10, 1].sum() / df.iloc[:, 1].sum())
print("81% of the master programs provided by just top 10 countries")
print("--"*50)


----------------------------------------------------------------------------------------------------
How many countries provides only one program ? 
              country_name  program_count
90                  Guyana              1
91                    Fiji              1
92                Ethiopia              1
93  Bosnia and Herzegovina              1
----------------------------------------------------------------------------------------------------
2 | How many programs provides by the top 5 countries?
     country_name  program_count
0   United States          24084
1  United Kingdom          14130
2       Australia           2932
3          Canada           2152
4     Netherlands           1346
0.8194455937112123
81% of the master programs provided by just top 10 countries
----------------------------------------------------------------------------------------------------


# 2 | which university gives the most master program

In [137]:
university_data.sort_values()

country_name
Armenia                              1
Uganda                               1
Trinidad and Tobago                  1
Tanzania, United Republic of         1
Palestinian Territory, Occupied      1
                                  ... 
Italy                               95
France                             152
Germany                            169
United Kingdom                     225
United States                      669
Name: university_name, Length: 94, dtype: int64

In [140]:
# Which country has most number of university ?

university_data = data.groupby(["country_name"])["university_name"].nunique().sort_values(ascending=False)
df = pd.DataFrame({"country_name": university_data.index,
                   "university_count": university_data.values})
fig = go.Figure()

fig.add_trace(
    go.Bar(x=df["country_name"], y = df["university_count"], marker_color= '#9900cc')
)

fig.update_layout(
    template="plotly_dark",
    title="The number of master programs provided by each countries.",
    yaxis_title="Number of universities",
    bargap=0.2,

)

#### The universities provide only one master program

In [146]:
[data["country_name"].isin(["Ethiopia", "Guyana", "Bosnia and Herzegovina", "Fiji"])]

[0        False
 1        False
 2        False
 3        False
 4        False
          ...  
 60420    False
 60421    False
 60422    False
 60423    False
 60424    False
 Name: country_name, Length: 60425, dtype: bool]

In [148]:
data[data["country_name"].isin(["Ethiopia", "Guyana", "Bosnia and Herzegovina", "Fiji"])]

country_name                  university_name  \
3888   Bosnia and Herzegovina           University of Sarajevo   
7786                 Ethiopia           Addis Ababa University   
7787                     Fiji  University of the South Pacific   
11191                  Guyana             University of Guyana   

       university_rank                       program_name program_type  \
3888               NaN                           Property       Master   
7786               NaN            Accounting and Auditing          MSc   
7787               NaN  Master of Business Administration          MBA   
11191              NaN                    Social Sciences       Master   

        duration language tution_1_currency  tution_1_money  tution_1_type  \
3888   24 months  English               EUR         10225.0  International   
7786   18 months  English               USD            80.0  International   
7787   12 months  English               EUR         12274.0  International   
11191  12 months  English               NaN             NaN            NaN   

      tution_2_currency  tution_2_money tution_2_type  \
3888                EUR          5100.0      National   
7786                USD            50.0      National   
7787                EUR          7651.0      National   
11191               NaN             NaN           NaN   

      tuition_price_specification  ielts_score             city  
3888               Tuition (Year)          NaN     ['Sarajevo']  
7786             Tuition (Credit)          NaN  ['Addis Ababa']  
7787               Tuition (Year)          NaN         ['Suva']  
11191                         NaN          NaN   ['Georgetown']

In [150]:
print(df.iloc[0:10, 1].sum() / df.iloc[:, 1].sum())
print("67% of universities which provide masters are present in only top 10 countries")
print("--"*50)

0.6770004189359028
67% of universities which provide masters are present in only top 10 countries
----------------------------------------------------------------------------------------------------


In [151]:
10/94

0.10638297872340426

- 67% of universities which provide masters are present in only top 10 countries. that is just 10.6% of countries (10/94).
So 67% universities are present in just 10.6% of countries.

In [156]:
university_data = data.groupby(["university_name"])["program_name"].count().sort_values(ascending=False)
df = pd.DataFrame({"university_name": university_data.index,
                   "program_count": university_data.values})
fig = go.Figure()

fig.add_trace(
    go.Bar(x=df["university_name"][:20], y = df["program_count"][:20], marker_color= '#9900cc')
)

fig.update_layout(
    template="plotly_dark",
    title="The number of master programs provided by each universities.",
    yaxis_title="Number of universities",
    bargap=0.2,

)

# 3 | Does country has any effects on the price from the same university?


In [163]:
university_data = data.groupby(["university_name"])["country_name"].nunique().sort_values(ascending=False)
university_country_count = pd.DataFrame({"university_name": university_data.index,
                   "num_of_countries_present_in": university_data.values})

print(f"Number of university {len(university_country_count)}")
university_country_count = university_country_count[university_country_count["num_of_countries_present_in"] > 1]
print(f"Number of university present in more than 1 countries  {len(university_country_count)}")


Number of university 2273
Number of university present in more than 1 countries  64


- we will select 5 universities which are present in different countries and check the available programs there

In [174]:
selected_university = university_country_count[:5]["university_name"].values
selected_uni_data = data[data["university_name"].isin(selected_university)]
selected_uni_data.groupby(["university_name"])["program_name"].count().reset_index()

university_name  program_name
0                EIT Digital Master School            72
1              ESCP Europe Business School           145
2      Engineering Institute of Technology            37
3  IHE Delft Institute for Water Education           115
4                       University of Kent          1161

In [189]:
selected_uni_data[(selected_uni_data["university_name"] == "EIT Digital Master School")]["program_name"]

7894               Cloud Computing and Services (CCS)
7895                                     Data Science
7896                   Digital Media Technology (DMT)
7897                            Embedded Systems (ES)
7898     Human Computer Interaction and Design (HCID)
                             ...                     
25778                           Embedded Systems (ES)
25779    Human Computer Interaction and Design (HCID)
25780      Internet Technology and Architecture (ITA)
25781                      Security and Privacy (S&P)
25782        Software and Service Architectures (SSA)
Name: program_name, Length: 72, dtype: object

In [179]:
# What are all the common programs in those 5 universities?
unique_program_name: list = selected_uni_data["program_name"].unique()



def select_program_present_in_all_university(df, programs):
    pass

In [188]:
selected_uni_data[(selected_uni_data["university_name"] == "EIT Digital Master School") & (selected_uni_data["program_name"] == "Cloud Computing and Services (CCS)")]

country_name            university_name  university_rank  \
7894          Finland  EIT Digital Master School              NaN   
8481           France  EIT Digital Master School              NaN   
9496          Germany  EIT Digital Master School              NaN   
11812         Hungary  EIT Digital Master School              NaN   
13801           Italy  EIT Digital Master School              NaN   
15461     Netherlands  EIT Digital Master School              NaN   
19688           Spain  EIT Digital Master School              NaN   
20511          Sweden  EIT Digital Master School              NaN   
25775  United Kingdom  EIT Digital Master School              NaN   

                             program_name program_type   duration language  \
7894   Cloud Computing and Services (CCS)          MSc  24 months  English   
8481   Cloud Computing and Services (CCS)          MSc  24 months  English   
9496   Cloud Computing and Services (CCS)          MSc  24 months  English   
11812  Cloud Computing and Services (CCS)          MSc  24 months  English   
13801  Cloud Computing and Services (CCS)          MSc  24 months  English   
15461  Cloud Computing and Services (CCS)          MSc  24 months  English   
19688  Cloud Computing and Services (CCS)          MSc  24 months  English   
20511  Cloud Computing and Services (CCS)          MSc  24 months  English   
25775  Cloud Computing and Services (CCS)          MSc  24 months  English   

      tution_1_currency  tution_1_money  tution_1_type tution_2_currency  \
7894                EUR          8000.0  International              Free   
8481                EUR          8000.0  International              Free   
9496                EUR          8000.0  International              Free   
11812               EUR          8000.0  International              Free   
13801               EUR          8000.0  International              Free   
15461               EUR          8000.0  International              Free   
19688               EUR          8000.0  International              Free   
20511               EUR          8000.0  International              Free   
25775               EUR          8000.0  International              Free   

       tution_2_money tution_2_type tuition_price_specification  ielts_score  \
7894              0.0        EU/EEA              Tuition (Year)          6.5   
8481              0.0        EU/EEA              Tuition (Year)          6.5   
9496              0.0        EU/EEA              Tuition (Year)          6.5   
11812             0.0        EU/EEA              Tuition (Year)          6.5   
13801             0.0        EU/EEA              Tuition (Year)          6.5   
15461             0.0        EU/EEA              Tuition (Year)          6.5   
19688             0.0        EU/EEA              Tuition (Year)          6.5   
20511             0.0        EU/EEA              Tuition (Year)          6.5   
25775             0.0        EU/EEA              Tuition (Year)          6.5   

                                                    city  
7894   ['Stockholm', 'Helsinki', 'Delft', 'Paris', 'B...  
8481   ['Stockholm', 'Helsinki', 'Delft', 'Paris', 'B...  
9496   ['Stockholm', 'Helsinki', 'Delft', 'Paris', 'B...  
11812  ['Stockholm', 'Helsinki', 'Delft', 'Paris', 'B...  
13801  ['Stockholm', 'Helsinki', 'Delft', 'Paris', 'B...  
15461  ['Stockholm', 'Helsinki', 'Delft', 'Paris', 'B...  
19688  ['Stockholm', 'Helsinki', 'Delft', 'Paris', 'B...  
20511  ['Stockholm', 'Helsinki', 'Delft', 'Paris', 'B...  
25775  ['Stockholm', 'Helsinki', 'Delft', 'Paris', 'B...

- The same university has the same fee in all countries as expected.